File: python/Algorithms/Sorting/insertion_sort.py
from typing import List
from Algorithms.Sorting.common import CompareFunction, Sortable, swap


def insertion_sort(input_list: List[Sortable],
                   comparator: CompareFunction) -> List[Sortable]:
    """"
    Sorts the input list, using the provided comparator
    to decide if two elements are in the correct order.
    Returns a copy of the list sorted, makes no changes to input list
    """
    output_list: List[Sortable] = list(input_list)

    for index in range(1, len(output_list)):
        item_to_place: Sortable = output_list[index]
        item_place: int = index

        while item_place > 0:
            lower: int = item_place - 1
            upper: int = item_place
            comparison: int = comparator(output_list[lower], item_to_place)

            # The comparator returns -1 if the first item is 'greater than' the second one
            if comparison < 0:
                # Temporary variable to prevent overwrites
                swap(output_list, lower, upper)
            else:
                # Item to place can be placed early
                item_place = upper
                break
            item_place -= 1

        output_list[item_place] = item_to_place

    return output_list

File: lib/src/algorithms/sort/insertionSort.ts
import { anyComparator, emptyObserver, simpleSwap } from "../../common";
import { SortUtility } from "./types";

function insertionSort<T>(
  inputList: T[],
  {
    compare = anyComparator,
    observe = emptyObserver,
    swap = simpleSwap,
  }: SortUtility<T>
): T[] {
  if (inputList.length < 2) {
    return inputList;
  }

  // Don't modify the input
  const outputList: T[] = [...inputList];

  for (let index = 1; index < outputList.length; index++) {
    observe("Placing Item", outputList, { index });
    let itemPlace: number = index;
    while (itemPlace > 0) {
      const lower: number = itemPlace - 1;
      const upper: number = itemPlace;

      observe("Seeking Place", outputList, {
        index,
        lower,
        upper,
      });

      const comparison: number = compare(outputList[lower], outputList[upper], {
        aIndex: lower,
        bIndex: upper,
      });

      // The compare returns -ve if the first item is 'greater than' the second one
      if (comparison > 0) {
        // Temporary variable to prevent overwrites
        swap(outputList, lower, upper);
      } else {
        itemPlace = upper;
        break;
      }

      itemPlace -= 1;
    }
  }

  return outputList;
}

export default insertionSort;
